libs/Microsoft.MixedReality.WebRTC.Native/src/interop/global_factory.cpp

```diff
@@ -54,61 +49,95 @@ std::string ObjectToString(ObjectType type, TrackedObject* obj) {
 
 namespace Microsoft::MixedReality::WebRTC {
 
-const std::unique_ptr<GlobalFactory>& GlobalFactory::Instance() {
-  return g_factory;
+uint32_t GlobalFactory::StaticReportLiveObjects() noexcept {
+  // Lock the instance to prevent shutdown if it already exists, while
+  // enumerating live objects.
+  RefPtr<GlobalFactory> factory(InstancePtrIfExist());
+  if (factory) {
+    return factory->ReportLiveObjects();
+  }
+  return 0;
 }
 
-GlobalFactory::~GlobalFactory() {
-  std::scoped_lock lock(mutex_);
-  if (!alive_objects_.empty()) {
-    // WebRTC object destructors are also dispatched to the signaling thread,
-    // like all method calls, but the threads are stopped by the GlobalFactory
-    // shutdown, so dispatching will never complete.
-    RTC_LOG(LS_ERROR) << "Shutting down the global factory while "
-                      << alive_objects_.size()
-                      << " objects are still alive. This will likely deadlock.";
-    for (auto&& pair : alive_objects_) {
-      RTC_LOG(LS_ERROR) << "- " << ObjectToString(pair.second, pair.first)
-                        << " [" << pair.first->GetApproxRefCount()
-                        << " ref(s)]";
-    }
+void GlobalFactory::SetShutdownOptions(mrsShutdownOptions options) noexcept {
+  // Unconditionally set shutdown options whether or not the instance is
+  // initialized. So no need to acquire the init lock.
+  GlobalFactory* const factory = GetInstance();
+  // However, acquire the thread-safety mutex protecting concurrent accesses.
+  std::scoped_lock lock(factory->mutex_);
+  factory->shutdown_options_ = options;
+}
+
+void GlobalFactory::ForceShutdown() noexcept {
+  GlobalFactory* const factory = GetInstance();
+  std::scoped_lock lock(factory->init_mutex_);
+  if (!factory->peer_factory_) {
+    return;
+  }
+  try {
+    factory->ShutdownImplNoLock(ShutdownAction::kForceShutdown);
+  } catch (...) {
   }
-  ShutdownNoLock();
 }
 
-rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface>
-GlobalFactory::GetOrCreate() {
-  std::scoped_lock lock(mutex_);
-  if (!factory_) {
-    if (Initialize() != Result::kSuccess) {
-      return nullptr;
-    }
+bool GlobalFactory::TryShutdown() noexcept {
+  GlobalFactory* const factory = GetInstance();
+  std::scoped_lock lock(factory->init_mutex_);
+  if (!factory->peer_factory_) {
+    return true;
+  }
+  try {
+    return factory->ShutdownImplNoLock(ShutdownAction::kTryShutdownIfSafe);
+  } catch (...) {
+    return false;  // failed to shutdown

```

Comment:
nit: should we try to print out what failed here?

Reply:
You mean in the catch() block? Not sure what we can print...

Reply:
Discussed offline, printing error message for `std::exception`, as WebRTC code shouldn't throw.